File: src/main/java/DataStructureAlgo/Java/LeetCode2025/ProblemSet/Graph/island/_695/MaxAreaOfIsland_695.java
package DataStructureAlgo.Java.LeetCode2025.ProblemSet.Graph.island._695;

import DataStructureAlgo.Java.LeetCode2025.ProblemSet.Graph.island._200.NumberOfIsland_200;
import DataStructureAlgo.Java.LeetCode2025.ProblemSet.Graph.island._305.NumberOfIslandII_305;
import DataStructureAlgo.Java.helpers.CommonMethods;

import java.util.HashMap;
import java.util.Map;
import java.util.stream.IntStream;

/**
 * Author: Nitin Gupta
 * Date: 9/13/2024
 * Question Category: 695. Max Area of Island
 * Description: https://leetcode.com/problems/max-area-of-island/description/
 * You are given an m x n binary matrix grid. An island is a group of 1's (representing land) connected 4-directionally (horizontal or vertical.) You may assume all four edges of the grid are surrounded by water.
 *
 * The area of an island is the number of cells with a value 1 in the island.
 *
 * Return the maximum area of an island in grid. If there is no island, return 0.
 *
 *
 *
 * Example 1:
 *
 *
 * Input: grid = [[0,0,1,0,0,0,0,1,0,0,0,0,0],[0,0,0,0,0,0,0,1,1,1,0,0,0],[0,1,1,0,1,0,0,0,0,0,0,0,0],[0,1,0,0,1,1,0,0,1,0,1,0,0],[0,1,0,0,1,1,0,0,1,1,1,0,0],[0,0,0,0,0,0,0,0,0,0,1,0,0],[0,0,0,0,0,0,0,1,1,1,0,0,0],[0,0,0,0,0,0,0,1,1,0,0,0,0]]
 * Output: 6
 * Explanation: The answer is not 11, because the island must be connected 4-directionally.
 * Example 2:
 *
 * Input: grid = [[0,0,0,0,0,0,0,0]]
 * Output: 0
 *
 *
 * Constraints:
 *
 * m == grid.length
 * n == grid[i].length
 * 1 <= m, n <= 50
 * grid[i][j] is either 0 or 1.
 * File reference
 * -----------
 * Duplicate {@link DataStructureAlgo.Java.LeetCode.island.MaxAreaIsland}
 * Similar {@link DataStructureAlgo.Java.companyWise.GroupOn.IslandsSizeCount}
 * extension {@link NumberOfIslandII_305}
 * <p><p>
 *
 * Tags
 * -----
 * @medium
 * @Array
 * @Depth-FirstSearch
 * @Breadth-FirstSearch
 * @UnionFind
 * @Matrix
 * @Graph
 *
 * <p><p>
 * Company Tags
 * -----
 * <p><p>
 * @Amazon
 * @Facebook
 * @Microsoft
 * @Google
 * @Qualtrics
 * @Adobe
 * @Affirm
 * @Alibaba
 * @Bloomberg
 * @ByteDance
 * @DoorDash
 * @Groupon
 * @Intuit
 * @LinkedIn
 * @Mathworks
 * @Oracle
 * @PalantirTechnologies
 * @Twitch
 * @Uber
 * @WalmartLabs
 * @Wish
 * @Zillow
 *
 *
 *
 * @Editorial https://leetcode.com/problems/max-area-of-island/solutions/380781/simple-easy-with-mxn-using-union-find
 */
public class MaxAreaOfIsland_695 {

    public static void main(String[] args) {
        boolean test = true;
        test &= test(new int[][]{{1,0,1}, {1,0,1}, {1,1,1,}}, 7);
        test &= test(new int[][]
                {
                        {0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0},
                        {0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0},
                        {0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0},
                        {0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0},
                        {0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0},
                        {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0},
                        {0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0},
                        {0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0}

                }, 6);


        test &= test(new int[][]{{1, 1, 1, 1, 1}, {1, 1, 0, 0, 0}, {0, 0, 0, 1, 1}, {0, 0, 0, 1, 1}}, 7);
        test &= test(new int[][]{{1, 1, 1, 1, 1}, {1, 1, 1, 1, 1}, {0, 0, 0, 1, 1}, {0, 0, 0, 1, 1}}, 14);
        test &= test(new int[][]{{0, 0, 0, 0, 0, 0, 0, 0}}, 0);

        CommonMethods.printAllTestOutCome(test);


    }


    private static boolean test(int[][] grid, int expected) {
        System.out.println("-------------------------------------");
        System.out.println("Grid:\n" + CommonMethods.toString(grid) + " expected :" + expected);

        Solution solution = new Solution();
        int maxAreaOfIsland = solution.maxAreaOfIsland(CommonMethods.copyOf(grid));
        boolean test = maxAreaOfIsland == expected;
        System.out.println("Obtained :" + maxAreaOfIsland + " test :" + (test?"passed":"failed"));

        SolutionUnionFind solutionUnionFind = new SolutionUnionFind();
        int maxAreaOfIslandUnionFind = solutionUnionFind.maxAreaOfIsland(CommonMethods.copyOf(grid));
        boolean testUnionFind = maxAreaOfIslandUnionFind == expected;
        System.out.println("Obtained UF:" + maxAreaOfIslandUnionFind + " test :" + (testUnionFind?"passed":"failed"));
        return test && testUnionFind;
    }


    static class Solution {
        public int maxAreaOfIsland(int[][] grid) {
            if (grid == null || grid.length == 0)
                return 0;

            int m = grid.length;
            int n = grid[0].length;
            int maxArea = 0;
            for(int i=0; i<m; i++) {
                for (int j = 0; j < n; j++) {

                    if(grid[i][j] == 1) {
                        maxArea = Math.max(maxArea, dfs(grid, i, j, m, n));
                    }
                }
            }
            return maxArea;
        }

        private int dfs(int[][] grid, int i, int j, int m, int n) {
            if(!isSafe(i, j, m, n) || grid[i][j] != 1) {
                return 0;
            }

            //mark visited
            grid[i][j] = 0;

            //try all 4 directions
            return 1 + dfs(grid, i + 1, j, m, n) +
                    dfs(grid, i - 1, j, m, n) +
                    dfs(grid, i, j + 1, m, n) +
                    dfs(grid, i, j - 1, m, n);

        }

        private boolean isSafe(int i, int j, int m, int n) {
            return i >= 0 && j >= 0 && i < m && j < n;
        }

    }

    /**
     * Similar {@link NumberOfIsland_200}
     * Using {@link DataStructureAlgo.Java.nonleetcode.UnionFindDisjointSets}
     *
     *
     * The idea is to create a graph of each connected land and get the size of that graph. Find a graph with the max size.
     * Using union Find, we can create a multi-component graph; each component in a graph would represent an island.
     * T/S: O(m*n * log mn) / O(mn)
     */
    static class SolutionUnionFind {

        int[] row = {0, 0, 1, -1};
        int[] col = {-1, 1, 0, 0};

        static class UnionFind {
            static class Set {
                int rank; //size
                int id;

                public Set(int id, int rank) {
                    this.rank = rank;
                    this.id = id;
                }

            }

            private final Set[] parents;
            Map<Integer,Integer> IdToIndex;
            int maxSize = 1;

            public UnionFind(int n) {
                //each element has size 1 and it is its own parent
                parents = IntStream.range(0, n).mapToObj(i -> new Set(i, 1)).toArray(Set[]::new);
            }

            //this will minimize the slots
            public UnionFind(int [][]grid) {
                int size = 0;

                //find slots that are required for only those, where is land
                IdToIndex = new HashMap<>();
                int m = grid.length;
                int n = grid[0].length;
                for(int i=0; i<m; i++) {
                    for(int j=0; j<n; j++) {
                        if(grid[i][j] == 1) {
                            int id = i * n + j;
                            IdToIndex.put(id, IdToIndex.size());
                        }
                    }
                }

                int totalLands = IdToIndex.size() + 1 ;

                //each element has size 1 and it is its own parent
                parents = IntStream.range(0, totalLands).mapToObj(i -> new Set(i, 1)).toArray(Set[]::new);
            }

            public int find(int i) {
                if (parents[i].id == i)
                    return i;
                return parents[i].id = find(parents[i].id); //path compression
            }

            public boolean union(int i, int j) {
                int ip = find(i);
                int jp = find(j);

                //both are already in the same set
                if (ip == jp)
                    return false;

                unionByRank(i, j, ip, jp);
                maxSize = Math.max(maxSize, Math.max(parents[jp].rank, parents[ip].rank));
                return true; //union was possible
            }

            private void unionByRank(int i, int j, int ip, int jp) {
                int ipRank = parents[ip].rank;
                int jpRank = parents[jp].rank;


                //graph of i as root is bigger as compare to j graph.
                if (ipRank > jpRank) {
                    parents[jp].id = ip; //attach i as parent of j, this makes size no difference in i's graph
                    parents[ip].rank += parents[jp].rank; //connected graph of i and j will be now has more nodes in it.
                }  else {
                    //both graph sizes are equal attached to i and  or j size is higher
                    parents[ip].id = jp; // make jp as parent of i, this makes j size graph increase by 1
                    parents[jp].rank += parents[ip].rank; //connected graph of i and j will be now has more nodes in it.
                }
            }

            public int getMaxSize() {
                return maxSize;
            }

        }



        public int maxAreaOfIsland(int[][] grid) {
            if(grid == null || grid.length == 0)
                return 0;

            int m = grid.length;
            int n = grid[0].length;

             UnionFind unionFind = new UnionFind(m*n);
//            UnionFind unionFind = new UnionFind(grid);
            boolean landFound = false;

            for (int i=0; i<m; i++) {
                for (int j=0; j<n; j++) {
                    if(grid[i][j] == 1){
                        landFound = true;


                        for(int k=0; k<4; k++) {
                            int r = i + row[k];
                            int c = j + col[k];
                            if(isSafe(r, c, m, n) && grid[r][c] == 1) {
                                int landIndex = _2DPos1DPost(i, j, m, n);
                                int neighbourIndex = _2DPos1DPost(r, c, m, n);
                                unionFind.union(landIndex, neighbourIndex);

                                /*
                                    int landIndex = _2DPos1DPost(i, j, m, n);
                                    int neighbourIndex = _2DPos1DPost(r, c, m, n);

                                    int landSlotIndex = unionFind.IdToIndex.get(landIndex);

                                    int neighbourSlotIndex = unionFind.IdToIndex.get(neighbourIndex);

                                    unionFind.union(landSlotIndex, neighbourSlotIndex);
                                 */
                            }
                        }
                    }
                }
            }
            if(!landFound)
                return 0;
            return unionFind.getMaxSize();

        }

        private boolean isSafe(int i, int j, int m, int n) {
            return i >= 0 && j >= 0 && i < m && j < n;
        }

        //translate 2D position to 1D index
        private int _2DPos1DPost(int i, int j, int m, int n) {
            //2d array has m rows and n columns, that makes it has m*n blocks.
            //if 'i' represent m, then i*n represent the position in 2D of row and addition, j makes column position
            return i * n + j;
        }
    }
}

File: src/main/java/DataStructureAlgo/Java/LeetCode2025/ProblemSet/Graph/island/_695/MaxAreaOfIsland_695.py
# Time / Space : O(n*m) / O(m*n)
from collections import deque
from typing import List


class Solution_DSU_Optimized:
    """Union find
    In a grid traversal starting from the top-left, you only actually need to check Right and Down. 
    Since the cells to the Left and Top have already been processed, Union-Find will naturally connect them. This cuts your inner loop operations in half!
    """

    class UnionFind:
        def __init__(self, n):
            self.parent = list(range(n))
            # we use union by size as we need to find the max_area which is nothing but size of 1's
            self.size = [1] * n
            self.max_size = 1

        def find(self, i) -> int:
            if self.parent[i] != i:
                self.parent[i] = self.find(self.parent[i])
            return self.parent[i]

        def union(self, i, j) -> bool:
            root_i = self.find(i)
            root_j = self.find(j)

            if root_i == root_j:
                return False

            size_i = self.size[root_i]
            size_j = self.size[root_j]

            if size_i >= size_j:
                self.parent[root_j] = root_i
                self.size[root_i] += self.size[root_j]
            else:
                self.parent[root_i] = root_j
                self.size[root_j] += self.size[root_i]

            self.max_size = max(self.max_size, self.size[root_i], self.size[root_j])

            return True

        def set_size(self):
            return self.max_size

    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])

        union_find = Solution.UnionFind(m * n)  # 2d to 1D
        max_area = 0

        # In a grid traversal starting from the top-left, you only actually need to check Right and Down. Since the cells to the Left and Top have already been processed, Union-Find will naturally connect them. This cuts your inner loop operations in half!
        direction = [[0, 1], [1, 0]]

        def _index(i, j):
            return i * n + j

        land_found = False
        max_size = 1
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    land_found = True
                    root = _index(i, j)
                    # run 4 dir

                    for dr, dc in direction:
                        r, c = dr + i, dc + j
                        if 0 <= r < m and 0 <= c < n and grid[r][c] == 1:
                            child = _index(r, c)
                            union_find.union(root, child)

        if land_found:
            return union_find.set_size()
        return 0


# Time / Space : O(n*m) / O(m*n)
class Solution_DSU:
    """Union find"""

    class UnionFind:
        def __init__(self, n):
            self.parent = list(range(n))
            # we use union by size as we need to find the max_area which is nothing but size of 1's
            self.size = [1] * n
            self.max_size = 1

        def find(self, i) -> int:
            if self.parent[i] == i:
                return i
            self.parent[i] = self.find(self.parent[i])
            return self.parent[i]

        def union(self, i, j) -> bool:
            root_i = self.find(i)
            root_j = self.find(j)

            if root_i == root_j:
                return False

            size_i = self.size[root_i]
            size_j = self.size[root_j]

            if size_i >= size_j:
                self.parent[root_j] = root_i
                self.size[root_i] += self.size[root_j]
            else:
                self.parent[root_i] = root_j
                self.size[root_j] += self.size[root_i]

            self.max_size = max(self.max_size, self.size[root_i], self.size[root_j])

            return True

        def set_size(self):
            return self.max_size

    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])

        union_find = Solution.UnionFind(m * n)  # 2d to 1D
        max_area = 0
        direction = [[1, 0], [0, 1], [-1, 0], [0, -1]]

        def _index(i, j):
            return i * n + j

        land_found = False
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    land_found = True
                    root = _index(i, j)
                    # run 4 dir
                    for dr, dc in direction:
                        r, c = dr + i, dc + j
                        if 0 <= r < m and 0 <= c < n and grid[r][c] == 1:
                            child = _index(r, c)
                            union_find.union(root, child)

        if land_found:
            return union_find.set_size()
        return 0


# Time / Space : O(n*m) / O(1)
class Solution_bfs:
    """ bfs """
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        queue = deque()
        m = len(grid)
        n = len(grid[0])
        max_area = 0
        direction = [[1,0], [0,1], [-1,0], [0,-1]]

        def bfs(i,j):
            queue.append((i,j))
            area = 1
            grid[i][j] = 0

            while queue:
                i,j = queue.popleft()
               
                for dir_i, dir_j in direction:
                    r , c = dir_i + i , dir_j + j
                    if 0 <= r < m and 0 <= c < n and grid[r][c] == 1:
                        queue.append((r,c))
                        area +=1 
                        grid[r][c] = 0
            
            return area


        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    max_area = max(max_area, bfs(i,j))
        
        return max_area

# Time / Space : O(n*m) / O(1)
class Solution_dfs:
    """ dfs """ 
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])
        max_area = 0

        def is_safe(i,j):
            return 0<=i<m and 0<=j<n and grid[i][j] == 1

        def dfs(i,j) -> int:

            if not is_safe(i,j):
                return 0
            
            grid[i][j] = 0

            return 1 + dfs(i+1,j) + dfs(i-1,j) + dfs(i, j+1) + dfs(i, j-1)
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    max_area = max(max_area, dfs(i,j))
                    if max_area == m * n:
                        return max_area
        
        return max_area